File: src/main/java/backend/academy/flame/MultiThreadFlame.java
package backend.academy.flame;

import backend.academy.exception.UnexpectedException;
import backend.academy.flame.task.ApplyGammaTask;
import backend.academy.flame.task.ComputeNormalsTask;
import backend.academy.function.Function;
import backend.academy.function.ListOfFunctions;
import backend.academy.pixel.Pixel;
import backend.academy.pixel.Pixels;
import backend.academy.variation.Variation;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicReference;

public class MultiThreadFlame extends Flame {
    private final ExecutorService threadPool = Executors.newCachedThreadPool();

    public MultiThreadFlame(ListOfFunctions listOfFunctions, List<Variation> variations, Pixels pixels) {
        super(listOfFunctions, variations, pixels);
    }

    @Override
    public void render(int sampleCount, int iterationPerSample, int symmetry) throws UnexpectedException {
        // Массив Future для хранения результатов задач
        Future<?>[] futures = new Future[sampleCount];

        // Запуск задач для каждой выборки
        for (int i = 0; i < sampleCount; i++) {
            futures[i] = threadPool.submit(() -> renderSample(iterationPerSample, symmetry));
        }

        // Ожидание завершения всех задач
        for (Future<?> future : futures) {
            try {
                future.get(); // Ждем завершения задачи
            } catch (InterruptedException | ExecutionException e) {
                throw new UnexpectedException(e.getMessage(), e);
            }
        }
        threadPool.shutdown();

        gammaCorrection();
    }

    @Override
    protected void updatePixel(int[] indexes, Function function) {
        var pixelRef = new AtomicReference<>(pixels.pixels()[indexes[0]][indexes[1]]);
        pixelRef.updateAndGet(currPixel -> {
            var newColor =
                currPixel.hitCount() == 0 ? function.startColor() : currPixel.color().mixWith(function.startColor());
            return new Pixel(newColor, currPixel.hitCount() + 1);
        });
        pixels.pixels()[indexes[0]][indexes[1]] = pixelRef.get();
    }

    @Override
    protected double computeNormals(double[][] normals) {
        try (var pool = ForkJoinPool.commonPool()) {
            pool.invoke(new ComputeNormalsTask(pixels.pixels(), normals));
        }
        return Arrays.stream(normals)
            .flatMapToDouble(Arrays::stream)
            .max()
            .orElse(0);

    }

    @Override
    protected void applyGammaCorrection(double[][] normals, double max) {
        try (var pool = ForkJoinPool.commonPool()) {
            pool.invoke(new ApplyGammaTask(pixels.pixels(), normals, GAMMA, max));
        }
    }

}

File: src/main/kotlin/backend/academy/flame/MultiThreadFlame.kt
package backend.academy.flame

import ApplyGammaTask
import backend.academy.exception.UnexpectedException
import backend.academy.backend.academy.flame.task.ComputeNormalsTask
import backend.academy.function.Function
import backend.academy.function.ListOfFunctions
import backend.academy.pixel.Pixel
import backend.academy.pixel.Pixels
import backend.academy.variation.Variation
import kotlinx.coroutines.*
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicReference

class MultiThreadFlame(listOfFunctions: ListOfFunctions?, variations: List<Variation?>?, pixels: Pixels?) :
    Flame(listOfFunctions, variations, pixels) {

    @Throws(UnexpectedException::class)
    override fun render(sampleCount: Int, iterationPerSample: Int, symmetry: Int) {

        runBlocking {
            val jobList = ArrayList<Job?>()
            repeat (sampleCount) {
                val job = launch {
                    renderSample(iterationPerSample, symmetry)
                }
                jobList.add(job)
            }
            jobList.forEach { it?.join() }
        }



        gammaCorrection()
    }

    override fun updatePixel(indexes: IntArray, function: Function) {
        val pixels1: Array<Array<Pixel>> = pixels.getPixels();
        val pixelRef = AtomicReference(pixels1[indexes[0]][indexes[1]])
        pixelRef.updateAndGet { currPixel: Pixel ->
            val newColor =
                if (currPixel.hitCount == 0) function.startColor else currPixel.color.mixWith(function.startColor)
            Pixel(newColor, currPixel.hitCount + 1)
        }
        pixels1[indexes[0]][indexes[1]]  = pixelRef.get()
    }

    override fun computeNormals(normals: Array<DoubleArray>): Double {
        return runBlocking {
            // Создаем задачу и выполняем её
            val task = ComputeNormalsTask(pixels.getPixels(), normals)
            return@runBlocking  task.compute()  // Здесь мы вызываем compute() для выполнения задачи
        }
    }

    override fun applyGammaCorrection(normals: Array<DoubleArray>, max: Double) {
        runBlocking {
            // Создаем задачу и выполняем её
            val task = ApplyGammaTask(pixels.getPixels(), normals, Flame.GAMMA,max)
            task.compute()  // Здесь мы вызываем compute() для выполнения задачи
        }
    }
}
